### ISSUE TITLE
mAP过低

### ISSUE THREAD

**Issue by lrxby** [VisionXLab/PointOBB-v3#2]
作者大大你好，我在DOTA-v1.0数据集上尝试了您项目中端到端和两阶段版本，但在24轮训练中，两者的mAP最高只有6%左右，端到端版本延长到120轮后，也只有10%，然后我下载了您的训练好的41.29%的.pth权重文件，但用它在数据集上测得的mAP也只有不到20%。但是数据集和原始标注是从官网下的，也是按照您的Readme文件中的过程处理的，配置文件也是用的您给出的，只更改了路径。请问这究竟是哪里的问题，是数据集的问题，还是配置文件的问题？
_base_ = [
    '../../configs/_base_/default_runtime.py'
]

norm_cfg = dict(type='GN', num_groups=32, requires_grad=True)  # add
debug = False

num_stages = 2
# num_stages = 1
model = dict(
    type='PointOBB',
    pretrained='torchvision://resnet50',
    backbone=dict(
        type='ResNet',
        depth=50,
        num_stages=4,
        out_indices=(0, 1, 2, 3),
        frozen_stages=1,
        norm_cfg=dict(type='BN', requires_grad=True),
        norm_eval=True,
        style='pytorch'),
    neck=dict(
        type='FPN',
        in_channels=[256, 512, 1024, 2048],
        out_channels=256,
        start_level=0,
        add_extra_convs='on_input',
        num_outs=4,
        norm_cfg=norm_cfg
    ),
    
    loss_diff_view=dict(type='SmoothL1Loss', beta=1.0, loss_weight=1.0),  # SSC loss
    crop_size = (1024, 1024),
    construct_view = True,   # rot/flp view
    construct_resize = True, # resized view
    weight_mode = 'dynamic_weight',
    scale_classes = [6, 8, 9, 11, 12],
    scale_classes2 = [0],
    
    roi_head=dict(
        type='PointOBBHead',
        num_stages=num_stages,
        top_k=7,
        with_atten=False,

        loss_symmetry_ss=dict(
            type='SmoothL1Loss', loss_weight=0.5, beta=0.1),
        angle_coder=dict(
                    type='PSCCoder',
                    angle_version='le90',
                    dual_freq=False,
                    num_step=3,
                    thr_mod=0),
        angle_version = 'le90',
        rotation_agnostic_classes=[1, 9, 11],
        agnostic_resize_classes = [2, 4],
        agnostic_resize_classes2 = [7],
        use_angle_loss = False, 
        add_angle_pred_begin = False, 
        not_use_rot_mil = False, 
        detach_angle_head = False,
        stacked_convs = 2,
        use_ssff = True,

        bbox_roi_extractor=dict(
            type='RotatedSingleRoIExtractor',
            roi_layer=dict(
                type='RoIAlignRotated',
                out_size=7,
                sample_num=2,
                clockwise=True),
            out_channels=256,
            featmap_strides=[4, 8, 16, 32]),
        bbox_head=dict(
            type='Shared2FCInstanceMILHead',
            num_stages=num_stages,
            with_loss_pseudo=False,
            in_channels=256,
            fc_out_channels=1024,
            roi_feat_size=7,
            num_classes=15,
            num_ref_fcs=0,
            bbox_coder=dict(
                type='DeltaXYWHBBoxCoder',
                target_means=[0., 0., 0., 0.],
                target_stds=[0.1, 0.1, 0.2, 0.2]),
            reg_class_agnostic=True,
            loss_type='MIL',
            loss_mil1=dict(
                type='MILLoss',
                binary_ins=False,
                loss_weight=0.25,
                loss_type='binary_cross_entropy'),
            loss_mil2=dict(
                type='MILLoss',
                binary_ins=False,
                loss_weight=0.25,
                loss_type='gfocal_loss'),),
    ),

    bbox_pred_head=dict(
        type='PointOBB_FCOS_Head',
        num_classes=15,
        in_channels=256,
        stacked_convs=4,
        feat_channels=256,
        strides=[4, 8, 16, 32],
        center_sampling=True,
        center_sample_radius=1.5,
        norm_on_bbox=True,
        centerness_on_reg=True,
        separate_angle=False,
        scale_angle=True,
        bbox_coder=dict(
            type='DistanceAnglePointCoder', angle_version='le90'),
        h_bbox_coder=dict(type='DistancePointBBoxCoder'),
        loss_cls=dict(
            type='FocalLoss',
            use_sigmoid=True,
            gamma=2.0,
            alpha=0.25,
            loss_weight=1.0),
        loss_bbox=dict(type='RotatedIoULoss', loss_weight=1.0),
        loss_angle=None,
        loss_centerness=dict(
            type='CrossEntropyLoss', use_sigmoid=True, loss_weight=1.0),
    ),

    # model training and testing settings
    train_cfg=dict(
        base_proposal=dict(
            base_scales=[4, 8, 16, 24, 32, 48, 64, 72, 80, 96],
            base_ratios=[1 / 3, 1 / 2, 1 / 1.5, 1.0, 1.5, 2.0, 3.0],
            shake_ratio=None,
            cut_mode='symmetry',  # 'clamp',
            gen_num_neg=0),
        fine_proposal=dict(
            gen_proposal_mode='fix_gen',
            cut_mode=None,
            shake_ratio=[0.1],
            base_ratios=[1, 1.2, 1.3, 0.8, 0.7],
            iou_thr=0.3,
            gen_num_neg=500,
        ),
        rcnn=None,
        iter_count = 0,  
        burn_in_steps1 = 16000, 
        burn_in_steps2 = 22000,  
    ),
    test_cfg=dict(
        rpn=None,
        rcnn=None,
        nms_pre=2000,
        min_bbox_size=0,
        score_thr=0.05,
        nms=dict(iou_thr=0.1),
        max_per_img=2000))

# dataset settings
#dataset_type = 'DOTAPointDataset'
dataset_type = 'DOTADataset' # for test
angle_version = 'le90'

img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(type='Resize', img_scale=(1024, 1024), keep_ratio=True),
    dict(type='RandomFlip', flip_ratio=0.5,version=angle_version) if not debug else dict(type='RandomFlip', flip_ratio=0.),

    dict(type='Normalize', **img_norm_cfg),
    dict(type='Pad', size_divisor=32),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels', 'gt_bboxes_ignore', 'gt_true_bboxes']),
]

test_pipeline = [
    dict(type='LoadImageFromFile'),
    #dict(type='LoadAnnotations', with_bbox=True),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(1024, 1024),
        flip=False,
        transforms=[
            dict(type='Resize',img_scale=(1024, 1024), keep_ratio=True),
            dict(type='Normalize', **img_norm_cfg),
            dict(type='Pad', size_divisor=32),
            dict(type='DefaultFormatBundle'),
            dict(type='Collect', keys=['img']),
        ])
]

data_root = '/media/chenjunjie/cdb0f1df-df9d-4704-ade1-ed1fde64603c/liurunxiang/dataset/DOTA1/'

data = dict(
    samples_per_gpu=2,
    workers_per_gpu=4,
    shuffle=False if debug else None,
    train=dict(
        type=dataset_type,
        version=angle_version,
        ann_file = data_root + "trainval2/trainval_1024_P2Bfmt_dotav10_rbox.json",
        img_prefix = data_root + 'trainval2/images/',
        pipeline=train_pipeline,
        filter_empty_gt=True
    ),
    val=dict(
        samples_per_gpu=8,
        type=dataset_type,
        ann_file = data_root + "trainval2/trainval_1024_P2Bfmt_dotav10_rbox.json",
        img_prefix = data_root + 'trainval2/images/',
        pipeline=test_pipeline,
        test_mode=False,
    ),
    test=dict(
        type=dataset_type,
        # ann_file=data_root + "test/images/",
        # img_prefix=data_root + 'test/images/',
        ann_file = data_root + "trainval2/trainval_1024_P2Bfmt_dotav10_rbox.json",
        img_prefix = data_root + 'trainval2/images/',        
        pipeline=test_pipeline))

check = dict(stop_while_nan=False)

# optimizer
optimizer = dict(type='SGD', lr=0.005, momentum=0.9, weight_decay=0.0001)

optimizer_config = dict(grad_clip=dict(max_norm=35, norm_type=2))
# learning policy
training_time = 2
lr_config = dict(
    policy='step',
    warmup='linear',
    warmup_iters=500,
    warmup_ratio=0.001,
    step=[8*training_time, 11*training_time])
runner = dict(type='EpochBasedRunner', max_epochs=12*training_time)

checkpoint_config = dict(interval=1)

**Comment by lrxby** [comment]
另外，我看您论文中41.29%的mAP是在测试集上的结果，但是官方并没有给测试机的标注，请问能分享你们用的数据集吗？

**Comment by Zzppppy** [comment]
你好，感谢你关注我们的工作！

DOTA系列数据集是比较特殊的，他公开了train,val和test集的数据集图片，以及train,val集的标注文件，而test集标注文件并未公开。
如果需要进行测试则要按照DOTA官方的教程，推理并生成结果zip文件然后上传至DOTA数据集的服务器进行评测，网页链接如下：
https://captain-whu.github.io/DOTA/

两阶段版本需要你推理出伪标签后自行用现有的mmrotate框架训练检测器然后进行测试；端到端版本可以按照我们readme中的指令进行推理并生成zip文件，然后上传DOTA服务器进行测试。

我们所有的mAP指标均是在test集上的结果。希望我的回答能帮助到你~

**Comment by lrxby** [comment]
非常感谢您的回复，在训练过程中每轮结束后都有一次评估，那请问在这个评估过程中端到端版本mAP最高大概是多少，我在24轮训练中最高只有6％左右，我延长到了120轮也只有10%左右，请问是正常的吗？如果不正常可能是什么问题呢？

**Comment by Zzppppy** [comment]
你好，你所说的指标的确属于异常情况。

我观察到你提供的config文件中修改了samples_per_gpu，这将与我们原来提供的优化器配置不匹配，从而导致训练出现异常情况。

请你将config文件中训练设定与我们提供的config文件保持一致。

**Comment by lrxby** [comment]
非常感谢您指出的错误，现在我有以下几个问题：
①如果samples_per_gpub保持原来的4，会报我的GPU内存不够的错误，我可以把samples_per_gpu改成2，然后把lr由原来的0.005改为0.0025吗？除此之外还需要做其他修改吗？
②如果我使用多张GPU训练，需要额外调整学习率lr吗，比如说默认是单卡训练，lr是0.005，在其他参数都不变的情况下如果我要用8张GPU训练的话lr是不是要改为原来的8倍？
③能介绍一下你们在训练过程中的mAP最高大概是多少吗，是和在测试集上的41.29%差不多吗？能让我在后续训练过程中能有参考，知道自己有没有出错误。
因为我是初学者，所以很抱歉提出这么多问题，有的可能还比较低级，万分感谢您在百忙之中能够回答我的问题。

**Comment by Zzppppy** [comment]
bs和lr最好成正比例进行修改。

如果你的bs扩大或缩小n倍，那么lr也可以尝试相应的扩大或缩小n倍。

mAP建议以测试集上的结果为准。
